master/buildbot/util/githubapiservice.py

```diff
@@ -0,0 +1,354 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+from __future__ import absolute_import
+from __future__ import print_function
+import datetime
+import pprint
+import textwrap
+import time
+from collections import namedtuple
+from collections import OrderedDict
+
+import jinja2
+import jwt
+from twisted.internet import defer
+from buildbot.util.httpclientservice import HTTPClientService
+from buildbot.util.logger import Logger
+from buildbot.util.service import SharedService
+
+log = Logger()
+
+InstallationAuthorization = namedtuple('InstallationAuthorization', ['token', 'expires_at', 'on_behalf_of'])
+
+
+class GithubIntegration(SharedService):
+
+    API_ROOT_URL = 'https://api.github.com'
+
+    def __init__(self, integration_id, private_key, api_root_url=None,
+                 debug=False, verify=None, **kwargs):
+        super(GithubIntegration, self).__init__()
+        self.integration_id = integration_id
+        self.private_key = private_key
+        self.api_root_url = api_root_url or self.API_ROOT_URL
+        self.debug = debug
+        self.verify = verify
+        self._tokens = {}
+        self._http = None
+
+    # Totally inspired in PyGithub https://goo.gl/xKXUtr
+
+    @defer.inlineCallbacks
+    def get_http_client(self):
+        if self._http is None:
+            self._http = yield HTTPClientService.getService(
+                self.master,
+                self.api_root_url,
+                headers={'User-Agent': 'Buildbot'},
+                debug=self.debug, verify=self.verify)
+        defer.returnValue(self._http)
+
+    def create_jwt(self):
+        """
+        Creates a signed JWT, valid for 60 seconds
+        """
+        now = int(time.time())
+        payload = {
+            "iat": now,
+            "exp": now + 60,
```

Comment:
That's a copy of PyGithub's library implementation, don't know why they had it like that. According to [GitHub's Documentation](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/about-authentication-options-for-github-apps/#authenticating-as-a-github-app) it cannot be longer than 10 minutes.